### DISCUSSION TITLE
Joint name mapping in HW interafec level (joint_1 -> joint_a1)

### DISCUSSION THREAD

**Discussion by farshbafdoustar** [ros-industrial/abb_robot_driver#78] [Q&A]
As I understood correctly the joints names in this package come from robot description got by RWS connection. As joint names are used in all interfaces exposed by ROS Control layer, is there any way to RE-NAME joint names for outside ROS world? So exchanging robots  (with different joint names defined in robot or even between different robot manufactures like exchanging ABB with KUKA) in HW interface layer does not need changing other layers (including URDF)?

If there is currently no solution, do you have any suggestion or preference that I keep in mind during implementation of this feature inside the HW interface by adding more data for example to egm related configuration files (params)?

**Comment by gavanderhoorn** [comment]
There is currently no support for renamed joints. Names are extracted from the system description returned by the IRC5 controller and processed for compliance with ROS naming conventions ([here](https://github.com/ros-industrial/abb_egm_rws_managers/blob/496597a50a0fc9a2cbee80498aa3bb6bc31b5908/src/system_data_parser.cpp#L620-L688)). `abb_robot_driver` treats the information the IRC5 provides as the truth.

> So exchanging robots (with different joint names defined in robot or even between different robot manufactures like exchanging ABB with KUKA) in HW interface layer does not need changing other layers (including URDF)?

the joint names in the `xacro:macro`s you refer to are wrong.

They should never have included the `a` character, and should have followed the convention to not include *any* OEM-specific information. So if anything would need fixing, it would be those `_macro.xacro`s.

See also [this section](https://gavanderhoorn.github.io/rep/rep-0199.html#link-n) in a document which describes something similar, but for `link`s (search for *rationale*).

> If there is currently no solution, do you have any suggestion or preference that I keep in mind during implementation of this feature inside the HW interface by adding more data for example to egm related configuration files (params)?

tbh I'm not entirely convinced yet of the necessity to change something in `abb_robot_driver`.

Perhaps if you could provide some more context and motivation, we could see what (if anything) needs to be done.

**Reply by farshbafdoustar** [reply]
> They should never have included the a character, and should have followed the convention to not include any OEM-specific information. So if anything would need fixing, it would be those _macro.xacros.

I'm completely agree with you on this point but I'm encountering this issue more and more when I switch between Robot HW interfaces. for example on [KUKA hw interface](https://github.com/ros-industrial/kuka_experimental/blob/melodic-devel/kuka_kr120_support/urdf/kr120r2500pro.urdf) the joints in URDF are defined based on the information communicated over RSI (joint_a1, joint a2,...). I'm agree on this point that URDF should be OEM convention agnostic but I think to keep it really independent from OEM internal conventions, we need a mapping somewhere.
For example in ROS2_control  , it is possible to utilize [actuator tag in hw description ](https://github.com/ros-controls/roadmap/blob/master/design_drafts/hardware_access.md) and adding some more extra OEM related information inside. But I'm wondering how we can handle in in ROS1 and what would be the best way from the ROS conventions point of view.
